esp-hal/src/dma/gdma.rs

```diff
@@ -30,541 +30,382 @@ use crate::{
-                            dma.in_int_ena_ch($num).modify(|_, w| w.in_suc_eof().clear_bit())
-                        } else {
-                            dma.int_ena_ch($num).modify(|_, w| w.in_suc_eof().clear_bit())
-                        }
-                    }
-                }
+    fn listen_ch_in_done() {
+        Self::in_int().ena().modify(|_, w| w.in_done().set_bit())
+    }
 
-                fn unlisten_out_eof() {
-                    let dma = unsafe { &*crate::peripherals::DMA::PTR };
-                    cfg_if::cfg_if! {
-                        if #[cfg(any(esp32c6, esp32h2, esp32s3))] {
-                            dma.out_int_ena_ch($num).modify(|_, w| w.out_total_eof().clear_bit())
-                        } else {
-                            dma.int_ena_ch($num).modify(|_, w| w.out_total_eof().clear_bit())
-                        }
-                    }
-                }
+    fn clear_ch_in_done() {
+        Self::in_int().clr().write(|w| w.in_done().set_bit());
+    }
 
-                fn listen_ch_in_done(){
-                    let dma = unsafe { &*crate::peripherals::DMA::PTR };
-                    cfg_if::cfg_if! {
-                        if #[cfg(any(esp32c6, esp32h2, esp32s3))] {
-                            dma.in_int_ena_ch($num).modify(|_, w| w.in_done().set_bit())
-                        } else {
-                            dma.int_ena_ch($num).modify(|_, w| w.in_done().set_bit())
-                        }
-                    }
-                }
+    fn is_ch_in_done_set() -> bool {
+        Self::in_int().raw().read().in_done().bit()
+    }
 
-                fn clear_ch_in_done(){
-                    let dma = unsafe { &*crate::peripherals::DMA::PTR };
+    fn unlisten_ch_in_done() {
+        Self::in_int().ena().modify(|_, w| w.in_done().clear_bit());
+    }
 
-                    #[cfg(not(any(esp32c6, esp32h2, esp32s3)))]
-                    dma.int_clr_ch($num).write(|w| w.in_done().set_bit());
-                    #[cfg(any(esp32c6, esp32h2, esp32s3))]
-                    dma.in_int_clr_ch($num).write(|w| w.in_done().set_bit());
-                }
+    fn is_listening_ch_in_done() -> bool {
+        Self::in_int().ena().read().in_done().bit()
+    }
+}
 
-                fn is_ch_in_done_set() -> bool {
-                    let dma = unsafe { &*crate::peripherals::DMA::PTR };
+#[non_exhaustive]
+pub struct ChannelTxImpl<const N: u8> {}
 
-                    #[cfg(not(any(esp32c6, esp32h2, esp32s3)))]
-                    let ret = dma.int_raw_ch($num).read().in_done().bit();
-                    #[cfg(any(esp32c6, esp32h2, esp32s3))]
-                    let ret = dma.in_int_raw_ch($num).read().in_done().bit();
+impl<const N: u8> TxChannel<Channel<N>> for ChannelTxImpl<N> {
+    #[cfg(feature = "async")]
+    fn waker() -> &'static embassy_sync::waitqueue::AtomicWaker {
+        static WAKER: embassy_sync::waitqueue::AtomicWaker =
```

Comment:
Just a post-mortem observation: This change is incorrect, as it only creates a single waker for Tx, and a single one for Rx, regardless of how many DMA channels exist. `static` variables act as though they were defined outside their containing item, so they don't know about any related generic parameters.